### DISCUSSION TITLE
Diff explorer shows "empty file" for non-zero length file with permissions change

### DISCUSSION THREAD

**Discussion by posita** [community/community#6814] [Pull Requests]
When I change the permissions of a file, but otherwise make no other changes, diffs claim the file is empty. From [this demo](https://github.com/posita/permissions-change-bug/pull/2/files):

<img width="438" alt="Screen Shot 2021-10-27 at 08 15 16" src="https://user-images.githubusercontent.com/222581/139073405-9404ced9-1eb0-48d5-8666-66d9d0480ada.png">

This is confusing. Users expect that to somehow indicate that no changes were made, not that the file is now empty. For consistency, users may even expect to be able to expand the file in place to see its (unchanged) contents like they can with other files.

<img width="25" alt="Screen Shot 2021-10-27 at 08 24 25" src="https://user-images.githubusercontent.com/222581/139074616-52cea3a4-e76a-491d-b0bf-efefd6d56f94.png">  👈 Note that the "expand all" icon is missing for files with contents whose permissions are changed with no other edits.

Note that the same behavior is present when viewing commits (see, e.g., [this](https://github.com/posita/permissions-change-bug/commit/11c4cf4d423a3bab5564f4f384783b134ae649f7) and [this](https://github.com/posita/permissions-change-bug/pull/2/commits/11c4cf4d423a3bab5564f4f384783b134ae649f7)).

A colleague who is new to GitHub brought this to my attention. He thought that meant I had deleted the contents of the file and pulled my branch down to examine it by hand. I realize he could have selected `View file` from the `…` menu, but he was unaware of that. Further, given the behavior of the other files in the diff (whose unchanged portions could be expanded in place), he didn't even think to look for that option. He was expecting to be able to expand it in place if it had contents, which is entirely reasonable. I'm kind of surprised this isn't already the behavior.

**Comment by chadcromwell** [comment]
This is not an answer, but I wanted to add a +1 to this.

It is quite confusing and has led to a bit of wasted time amongst our developers. Not sure if it is showing due to permission changes on our end or not.

It should follow the flow/behaviour of everything else. Or, give a better explanation of the change instead of "Empty file."

**Comment by posita** [comment]
[Anyone? Anyone?](https://youtu.be/uhiCFdWeQfA)

**Comment by danielcettour** [comment]
I faced this problem after changing the permissions of my files locally. I just reverted that. Following this instructions: [https://gist.github.com/liitfr/86c77bf0ea4a627ab9b62ad30c32c251](https://gist.github.com/liitfr/86c77bf0ea4a627ab9b62ad30c32c251)
If you are using ubuntu, the xargs command will not work with the sintaxis shown there. You should run it like this:
`find ~ -type f -print0 | xargs -0 chmod -x`
Push the new changes and those "Empty files" will be gone

**Reply by posita** [reply]
@danielcettour, I think this is a _non sequitur_. The gist you reference instructs users to set `git` to [ignore permissions](https://git-scm.com/docs/git-config#Documentation/git-config.txt-corefileMode) for a particular repository (which, as far as I know, would have to be re-introduced manually by each user for each clone of a repository). For many, many repositories, permissions are important. for example, build or translation scripts require executable permissions to be functional upon checkout. However, it is well accepted that blanketing entire file trees with executable permissions presents unnecessary risks. My original post highlights a very confusing aspect of GitHub's UX, which is independent of `git`. The advice to configure `git` to ignore permissions seems like a really blunt and odd work-around to avoid triggering GitHub's UX problem.

**Comment by thomasdane** [comment]
The solution for me was to make some small change to the file (e.g. add whitespace) and the push. The entire diff now shows again. You can revert the whitespace addition if needed.

**Reply by posita** [reply]
That's not a solution. It requires that developers know about and proactively apply non substantive edits specifically to work around GitHub's UX bug. The UX is misleading and should be addressed.

**Comment by coalsont** [comment]
How is this bug still unaddressed after 3 years?  The page logic just needs to know whether what is being rendered is a diff or a file, and change the message to `Contents unchanged.` or `No edits made.` instead of the scary and unhelpful `Empty file.` in that case.

**Comment by sivaramkjs** [comment]
I just found this thread while trying to resolve the same problem. I was able to fix it using the below command based on "ChatGPT" suggestion. Hence, posting the followed steps here for reference just in case if anyone stumbles on the same issue again in future.

1. Executed this command only for the file with issue: `chmod +x <file-path>`
2. The file was then showed up as modified with below output from `git diff` command:
```
diff --git a/<file-path> b/<file-path>
old mode 100644
new mode 100755
```
3. Committed and pushed the change to the repo. 
4. The `Empty file 100755 → 100644` diff was removed from the pull request changes on GitHub UI.

**Comment by bcieszko** [comment]
2025, still exists - it should be relatively simple fix on gh side..
